### PR TITLE
Use correct path for NewtonsoftJson ModelState errors

### LINKED ISSUE #33451: ModelState errors reported by Microsoft.AspNetCore.Mvc.NewtonsoftJson
### Describe the bug
Problem is related to the migration from ASP.NET MVC (.NET Framework 4.8) web application to .NET 5 and is about different amount of errors when parsing wrong JSON request.

In old code we had some JsonFormatter configuration
```csharp
public void RegisterConfiguration(HttpConfiguration config, IDependenciesResolver resolver)
{
    // ... exact settings don't matter here - just an example
    config.Formatters.JsonFormatter.SerializerSettings = new JsonSerializerSettings
        { NullValueHandling = NullValueHandling.Ignore };
    config.Formatters.JsonFormatter.SerializerSettings.Converters.Add(new StringEnumConverter());
    // ...
```
We have migrated the app and moved to Microsoft.AspNetCore.Mvc.NewtonsoftJson so our settings changed to
```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddControllers(options =>
    {
        // ...
    })
    .AddNewtonsoftJson(options =>
    {
        // example setting - for the sake of issue there can be no options set
        options.AllowInputFormatterExceptionMessages = false;
    });
```
We have some API POST endpoint that consumes some data model parsed from json, example:
```json
{
  "surpriseTime": "2021-06-10T22:34:51.702Z",
  "details": {
    "candyCount": abc
  }
}
```
As you can notice, for "candyCount" there is an incorrect value assigned.
For old app we received error message
```json
[
  {
    "Message": "Unexpected character encountered while parsing value: a. Path 'Details.CandyCount', line 4, position 19.",
    "Property": "surpriseModel.Details.CandyCount"  // this is a Key of ModelState KeyValuePair entry.
  }
]
```
In migrated app we get two errors
```json
[
  {
    "message": "Unexpected character encountered while parsing value: a. Path 'details.candyCount', line 4, position 19.",
    "property": "details.candyCount"
  },
  {
    "message": "Unexpected character encountered while parsing value: b. Path 'details.candyCount', line 4, position 19.",
    "property": "details.candyCount.details"
  }
]
```
If I get rid of `AddNewtonsoftJson` call, there is one message again.

I have checked the behavior of Newtonsoft.Json for both frameworks by attaching to SerializerSettings.Error delegate and in both cases the same amount of calls is done - it is 3 for such json as above.
If I nest some objects then the number of errors grows by 1 for each level.
There is also some weird property path generated for the errors. While first one does make sense, second one does not as there is no "details" property in type behind the "candyCount" (it is `int`). Type of the property does not matter - for DateTime that I tested error messages were the same.

I'm not sure if this is more a bug or a feature - Microsoft.AspNetCore.Mvc.NewtonsoftJson is more exact about number of errors coming from Newtonsoft.Json. On the other hand it does not make to much sense for me to have same multiple errors for one field. There is also the topic of weird key for ModelState entries.

### To Reproduce
Sample code is provided here https://github.com/wiesniak/NewtonsoftJsonIssue
There are two similar projects, one for .NET 5 and one for .NET Framework 4.8.
When run, swagger should load with one endpoint. Put the json payload as mentioned above and execute the call. Response should contain errors as described.
Now about the code. There is a class `ModelStateValidationAttribute` (one for each project) that is a global action filter with `OnActionExecuting` method override. If you put a breakpoint there and investigate the `actionContext.ModelState` you'll see the amount of errors and should notice the difference between projects.

#### dotnet --info
```
.NET SDK (reflecting any global.json):
 Version:   5.0.203
 Commit:    383637d63f

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.203\

Host (useful for support):
  Version: 5.0.6
  Commit:  478b2f8c0e

.NET SDKs installed:
  2.1.524 [C:\Program Files\dotnet\sdk]
  3.1.409 [C:\Program Files\dotnet\sdk]
  5.0.104 [C:\Program Files\dotnet\sdk]
  5.0.203 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

Code was executed via Visual Studio 2019 (16.9.6) but was also observed when app was deployed on Azure Web App.

### PR DESCRIPTION
**Pull request by halter73**
# Use correct path for NewtonsoftJson ModelState errors

The NewtonsoftJsonInputFormatter has logic that's intended to append the names of missing required properties to the ModelStateDictionary key. Normally, just the ModelName and ErrorContext.Path is used for this key, but ErrorContext.Path does not include the missing required property name like we want it to. For example, given the following class and input missing the required "Name" property:

```C#
class Person
{
    [JsonProperty(Required = Required.Always)]
    public string Name { get; set; }
}
```

We will see the following ErrorContext:

```
Error   {"Required property 'Name' not found in JSON. Path 'Person'..."} System.Exception {Newtonsoft.Json.JsonSerializationException}
Member  "Name"   object {string}
Path    "Person" string
```

So the path used for the ModelStateDictionary key is updated to be "Person.Name" instead of just "Person".
See https://github.com/aspnet/Mvc/issues/8509

However, this logic is caused problems with different JSON deserilization errors that include the member in the path but not at the end. For example, given the following classes and invalid input like `{ "b": { "c": { "d": abc } } }`:

```C#
class A
{
    public B B { get; set; }
}
class B 
{
    public C C { get; set; }
}
class C 
{
    public string D { get; set; }
}
```

We will see the following ErrorContext:

```
Error   {"Unexpected character encountered while parsing value: b. Path 'b.c.d'..."} System.Exception {Newtonsoft.Json.JsonReaderException}
Member  "c"     object {string}
Path    "b.c.d" string
```

Notice that Member "c" is in the middle of the Path "b.c.d". The error handler gets invoked for each level of nesting. null, "b", "c" and "d" are each a Member in different ErrorContexts all reporting the same parsing error.

The parsing error is reported as a JsonReaderException instead of as a JsonSerializationException like for missing required properties. We use the exception type to filter out these errors and keep the path used for the ModelStateDictionary key as "b.c.d" instead of "b.c.d.c" like before.

Fixes #33451

### CODE REVIEW

File: src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs [hunk 1]
```diff
@@ -211,6 +206,12 @@ public override Task JsonFormatter_EscapedKeys_Bracket()
         return base.JsonFormatter_EscapedKeys_Bracket();
     }
 
+    [Fact(Skip = "Expected: [0]['It\\'s a key'], Actual: [0]['It\\'s a key'].It's a key")]
```
Comment (halter73) [inline src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs:209]: I discovered this issue while working on this. I'll resolve this or file another issue and reference it here before merging the PR.

`ErrorContext.Path` escapes the `'`, but `ErrorContext.Member` doesn't meaning the `!path.EndsWith()` checks don't cover this scenario. We could scan `ErrorContext.Member` for any `'` characters and manually escape it before doing the `!path.EndsWith()` checks, but that feels like playing whac-a-mole. I'm guessing there are even more edge cases the `!path.EndsWith()` checks don't cover.

I really want to get rid of the `addMember` logic altogether and have Json.NET either give us the path we want to begin with or expose an `ErrorContext.ErrorType` so we only do the `addMember` logic for missing required properties where we know the member is never in the path. 

This is what @JamesNK suggested on @dougbu's PR a while back, but it looks like it wasn't possible due to time constraints. See https://github.com/aspnet/Mvc/pull/8659#discussion_r229449516 Is this something we could do for .NET 7 @JamesNK?

File: src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs [hunk 2]
```diff
@@ -234,30 +234,81 @@ void ErrorHandler(object? sender, Newtonsoft.Json.Serialization.ErrorEventArgs e
...
+            // Path    "b.c.d" string
+            //
+            // Notice that Member "c" is in the middle of the Path "b.c.d". The error handler gets invoked for each level of nesting.
+            // null, "b", "c" and "d" are each a Member in different ErrorContexts all reporting the same parsing error.
+            //
+            // The parsing error is reported as a JsonReaderException instead of as a JsonSerializationException like
+            // for missing required properties. We use the exception type to filter out these errors and keep the path used
+            // for the ModelStateDictionary key as "b.c.d" instead of "b.c.d.c"
+            // See https://github.com/dotnet/aspnetcore/issues/33451 
+            var addMember = !string.IsNullOrEmpty(member) && eventArgs.ErrorContext.Error is JsonSerializationException;
```
Comment (halter73) [inline src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs:288]: This additional `&& eventArgs.ErrorContext.Error is JsonSerializationException` check is the key to the fix.

Comment (captainsafia) [reply src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs:288]: Is `JsonSerializationException` the only exception type other than `JsonReaderException` that will surface in the error context? AKA, is this going to inadvertently break the way the member path is reported for other exceptions?

File: src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs (hunk 1 above)
Comment (JamesNK) [reply src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs:209]: I remember that issue. It was rather hacky.

The improvement I mentioned needs the discussed feature added to Newtonsoft.Json. I haven't thought any more about adding it.

File: src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs (hunk 2 above)
Comment (halter73) [reply src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs:288]: This is why I dug up the original PR that added this logic. As far as I can tell, this logic was added only to better report errors for missing required properties. If this happened to improve the path for other exception types, that was an unintended side effect which there are no tests for.

This and other issues like those involving escaped keys shows that the `addMember` logic is prone to adding the member to the path when it shouldn't, so my goal was to append the member to the path in the fewest cases possible without regressing the missing required property scenario.

**Comment by halter73** [comment]
I created a new issue for the single quote escaping and linked to it from the skipped test. I hoping to get this approved today since I'm OOF starting tomorrow.

File: src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs (hunk 1 above)
Comment (dougbu) [reply src/Mvc/Mvc.NewtonsoftJson/test/NewtonsoftJsonInputFormatterTest.cs]: > This is what @JamesNK suggested on @dougbu's PR a while back, but it looks like it wasn't possible due to time constraints. See https://github.com/aspnet/Mvc/pull/8659#discussion_r229449516 Is this something we could do for .NET 7 @JamesNK?

nit: The problem wasn't exactly time constraints; it was the lack of the feature @JamesNK mentioned (then and now).

> I remember that issue. It was rather hacky.

My bad. This fix is a bit more complicated and (unfortunately) not something I thought of last time around.

**Review by pranavkm — APPROVED** [review]
Thanks for the very detailed PR description!

**Pull request merged**